py/picca/pk1d/postproc_pk1d.py

```diff
@@ -450,27 +816,45 @@ def run_postproc_pk1d(data_dir,
     """
 
     if os.path.exists(output_file) and not overwrite:
-        raise RuntimeError('Output file already exists: ' + output_file)
+        raise RuntimeError("Output file already exists: " + output_file)
 
-    searchstr = '*'
+    searchstr = "*"
     files = glob.glob(os.path.join(data_dir, f"Pk1D{searchstr}.fits.gz"))
 
     with Pool(ncpu) as pool:
         output_readpk1d = pool.starmap(
-            read_pk1d, [[f, kbin_edges, snrcut, zbins_snrcut] for f in files])
+            read_pk1d, [[f, kbin_edges, snrcut, zbins_snrcut] for f in files]
+        )
 
     output_readpk1d = [x for x in output_readpk1d if x is not None]
-    p1d_table = vstack(
-        [output_readpk1d[i][0] for i in range(len(output_readpk1d))])
+    p1d_table = vstack([output_readpk1d[i][0] for i in range(len(output_readpk1d))])
     z_array = np.concatenate(
-        tuple(output_readpk1d[i][1] for i in range(len(output_readpk1d))))
-
-    userprint('Individual P1Ds read, now computing statistics.')
-    mean_p1d_table, metadata_table = compute_mean_pk1d(p1d_table, z_array,
-                                                      zbin_edges, kbin_edges,
-                                                      weight_method, apply_z_weights, nomedians,
-                                                      velunits, output_snrfit)
-    result = fitsio.FITS(output_file, 'rw', clobber=True)
-    result.write(mean_p1d_table.as_array(), header={'velunits': velunits})
-    result.write(metadata_table.as_array())
+        tuple(output_readpk1d[i][1] for i in range(len(output_readpk1d)))
+    )
+
+    userprint("Individual P1Ds read, now computing statistics.")
+
+    mean_p1d_table, metadata_table, cov_table = compute_mean_pk1d(
+        p1d_table,
+        z_array,
+        zbin_edges,
+        kbin_edges,
+        weight_method,
+        apply_z_weights,
+        nomedians=nomedians,
+        velunits=velunits,
+        output_snrfit=output_snrfit,
+        compute_covariance=compute_covariance,
+        compute_bootstrap=compute_bootstrap,
+        number_bootstrap=number_bootstrap,
+    )
+
+    result = fitsio.FITS(output_file, "rw", clobber=True)
+    result.write(mean_p1d_table.as_array())
+    result.write(
+        metadata_table.as_array(),
+        header={"VELUNITS": velunits, "NQSO": len(np.unique(p1d_table["forest_id"]))},
+    )
+    if cov_table is not None:
+        result.write(cov_table.as_array())
     result.close()
```

Comment:
@Waelthus I would naturally have stored the covmat as an image HDU (more precisely a 3D array), indeed with covmat.shape = (nz, nk, nk). Would it be ok to do that?

Reply:
About your point on Table.write: I think I tried this and then had issue with the custom header (does the option "meta" do the same as FITS's header?). Can be tested again.

Reply:
in principle it should do write the dict inside meta into the header, not that I'm against the current implementation. Just wanted to note it could be simpler.

Regarding storage: I would do as you say store it in a (nz, nk, nk) structure as long as we're not considering nk being dependent on z, if we want to implement that (and we might need to) then it indeed makes more sense to do a nz*nk(z)*nk(z) structure, but also the whole algorithm would become more complicated